substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/events/ThreadCPULoadEvent.java

```diff
@@ -0,0 +1,167 @@
+    }
+
+    @Uninterruptible(reason = "Accesses a JFR buffer.")
+    private static void emitForThread(IsolateThread isolateThread) {
+
+        long currCpuTime = getThreadCpuTime(isolateThread, true);
+        long prevCpuTime = JfrThreadLocal.getCpuTime(isolateThread);
+
+        long currWallClockTime = getWallClockTime();
+        long prevWallClockTime = JfrThreadLocal.getWallClockTime(isolateThread);
+        JfrThreadLocal.setWallClockTime(isolateThread, currWallClockTime);
+
+        // Threshold of 1 ms
+        if (currCpuTime - prevCpuTime < 1 * NANOSECS_PER_MILLISEC) {
+            return;
+        }
+
+        long currUserTime = getThreadCpuTime(isolateThread, false);
+        long prevUserTime = JfrThreadLocal.getUserTime(isolateThread);
+
+        long currSystemTime = currCpuTime - currUserTime;
+        long prevSystemTime = prevCpuTime - prevUserTime;
+
+        // The user and total cpu usage clocks can have different resolutions, which can
+        // make us see decreasing system time. Ensure time doesn't go backwards.
+        if (prevSystemTime > currSystemTime) {
+            currCpuTime += prevSystemTime - currSystemTime;
+            currSystemTime = prevSystemTime;
+        }
+
+        int processorsCount = getProcessorCount();
+
+        long userTime = currUserTime - prevUserTime;
+        long systemTime = currSystemTime - prevSystemTime;
+        long wallClockTime = currWallClockTime - prevWallClockTime;
+        float totalAvailableTime = wallClockTime * processorsCount;
+
+        // Avoid reporting percentages above the theoretical max
+        if (userTime + systemTime > wallClockTime) {
+            long excess = userTime + systemTime - wallClockTime;
+            currCpuTime -= excess;
+            if (userTime > excess) {
+                userTime -= excess;
+                currUserTime -= excess;
+            } else {
+                excess -= userTime;
+                currUserTime -= userTime;
+                userTime = 0;
+                systemTime -= excess;
+            }
+        }
+
+        JfrThreadLocal.setCpuTime(isolateThread, currCpuTime);
+        JfrThreadLocal.setUserTime(isolateThread, currUserTime);
+
+        JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
+        JfrNativeEventWriterDataAccess.initializeThreadLocalNativeBuffer(data);
+
+        JfrNativeEventWriter.beginSmallEvent(data, JfrEvent.ThreadCPULoad);
+        JfrNativeEventWriter.putLong(data, JfrTicks.elapsedTicks());
+        JfrNativeEventWriter.putEventThread(data);
+        JfrNativeEventWriter.putFloat(data, userTime / totalAvailableTime);
+        JfrNativeEventWriter.putFloat(data, systemTime / totalAvailableTime);
+        JfrNativeEventWriter.endSmallEvent(data);
+    }
+
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
+    private static int getProcessorCount() {
+        int currProcessorCount = Jvm.JVM_ActiveProcessorCount();
```

Comment:
JDK [JfrThreadCPULoadEvent::get_processor_count()](https://github.com/openjdk/jdk/blob/5d9baa2f9385d66cdf2ff8f065d165385eb48a22/src/hotspot/share/jfr/periodic/jfrThreadCPULoadEvent.cpp#L42) method calls `os::active_processor_count()` which includes fixes on linux:
[8140793 getAvailableProcessors may incorrectly report the number of cpus in Docker container](https://bugs.openjdk.org/browse/JDK-8140793)
[6515172 Runtime.availableProcessors() ignores Linux taskset command](https://bugs.openjdk.org/browse/JDK-6515172)

It looks like the `Containers.activeProcessorCount()` needs to be used to emit `ThreadCPULoad` event.

What I see the `Containers.activeProcessorCount()` is not uninterruptible:
https://github.com/oracle/graal/blob/4e0467860192075202641502c52c16747dc43721/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/Containers.java#L67

What is the right way to call interruptible `Containers.activeProcessorCount()` method from an uninterruptible code?

Reply:
Maybe it would be good enough to call and cache the result of `Containers.activeProcessorCount()` from interruptible code before you enqueue the VMoperation. I think this should be fine because the processor count you get will still be with respect to the interval of time you are calculating the thread CPU load over. The difference is that you poll somewhere in the middle of the interval instead of at the end.

Reply:
This should solve the case where the event is emitted from the interruptible `EndChunkNativePeriodicEvents.emit()` path.
It is still not clear for me what can be done with the case where the event is emitted from the pure uninterruptible `JfrThreadLocal.afterThreadExit(IsolateThread, Thread)` method.
https://github.com/oracle/graal/blob/16913ce1f8b3849cf371d2c0d8acbf345590482a/substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrThreadLocal.java#L134

Reply:
Unfortunately, `Containers.activeProcessorCount()` is currently not implemented in a way that it can be called from uninterruptible code. We will need to change that at some point. For now, I added a comment.